pkg/virt-handler/node-labeller/cpu_plugin.go

```diff
@@ -0,0 +1,196 @@
+			continue
+		}
+
+		newFeatures = subtractMap(n.cpuInfo.models[model], basicFeaturesMap)
+		features = unionMap(features, newFeatures)
+
+		cpus = append(cpus, model)
+	}
+	return cpus, features
+}
+
+//loadHostCapabilities loads info about cpu models, which can host emulate
+func (n *NodeLabeller) loadHostCapabilities() error {
+	hostDomCapabilities, err := n.virshCommandFN()
+	if err != nil {
+		return err
+	}
+
+	usableModels := make(map[string]bool)
+	for _, mode := range hostDomCapabilities.CPU.Mode {
+		for _, model := range mode.Model {
+			if model.Usable == usableNo || model.Usable == "" {
+				continue
+			}
+			usableModels[model.Name] = true
+		}
+	}
+	n.hostCapabilities.items = usableModels
+	return nil
+}
+
+//loadCPUInfo load info about all cpu models
+func (n *NodeLabeller) loadCPUInfo() error {
+	files, err := ioutil.ReadDir("/usr/share/libvirt/cpu_map/")
+	if err != nil {
+		return err
+	}
+
+	models := make(map[string]map[string]bool)
+	for _, f := range files {
+		fileName := f.Name()
+		if strings.HasPrefix(fileName, "x86_") {
+			features, err := loadFeatures(fileName)
+			if err != nil {
+				return err
+			}
+			cpuName := strings.TrimRight(strings.TrimLeft(fileName, "x86_"), ".xml")
+
+			models[cpuName] = features
+		}
+	}
+	n.cpuInfo.models = models
+
+	return nil
+}
+
+func runVirshCommand() (HostDomCapabilities, error) {
+	out, err := exec.Command("virsh", "domcapabilities", "--machine", "q35", "--arch", "x86_64", "--virttype", "kvm").Output()
+	if err != nil {
+		return HostDomCapabilities{}, err
+	}
+
+	hostDomCapabilities := HostDomCapabilities{}
+	err = xml.Unmarshal(out, &hostDomCapabilities)
+
+	return hostDomCapabilities, err
+}
+
+//LoadFeatures loads features for given cpu name
+func loadFeatures(fileName string) (map[string]bool, error) {
+	if fileName == "" {
+		return map[string]bool{}, nil
+	}
+
+	cpuFeatures := FeatureModel{}
+	cpuFeaturepath := getPathCPUFefatures(fileName)
+	err := getStructureFromXMLFile(cpuFeaturepath, &cpuFeatures)
+	if err != nil {
+		return nil, err
+	}
+
+	features := make(map[string]bool)
+	for _, f := range cpuFeatures.Model.Features {
+		features[f.Name] = true
+	}
+	return features, nil
+}
+
+//getPathCPUFefatures creates path where folder with cpu models is
+func getPathCPUFefatures(name string) string {
+	return "/usr/share/libvirt/cpu_map/" + name
+}
+
+func unionMap(a, b map[string]bool) map[string]bool {
+	unionMap := make(map[string]bool)
+	for feature := range a {
+		unionMap[feature] = true
+	}
+	for feature := range b {
+		unionMap[feature] = true
+	}
+	return unionMap
+}
+
+func subtractMap(a, b map[string]bool) map[string]bool {
+	new := make(map[string]bool)
+	for k := range a {
+		if _, ok := b[k]; !ok {
+			new[k] = true
+		}
+	}
+	return new
+}
+
+//GetStructureFromXMLFile load data from xml file and unmarshals them into given structure
+//Given structure has to be pointer
+func getStructureFromXMLFile(path string, structure interface{}) error {
```

Comment:
This function is only called from one place, so it can be more specific. It can load `FeatureModel` directly and return `(&FeatureModel, error)`.